scheduled_video_capture/scheduled_video_capture.py

```diff
@@ -0,0 +1,277 @@
+                return ""
+        
+            #Write all frames
+            for i, frame in enumerate(frames):
+                try:
+                    writer.write(frame)
+                except Exception as e:
+                    print(f"ERROR: Failed to write frame {i}: {str(e)}")
+        
+            #Release resources
+            writer.release()
+        
+            #Clean up
+            result_path = file_path
+            del self.sessions[camera_id][task_uuid]
+        
+            return result_path
+
+        return ""
+
+def get_quality_perc(resolution):
+    if resolution == 'Original':
+        return 100
+    elif resolution == 'Lower':
+        return 75
+    return 50
+
+#Create a global session manager
+capture_sessions = VideoCaptureSessions()
+
+def save_frame(frame, camera_id, task_uuid, task_settings, task_id):    
+    #Make a deep copy of the frame to avoid reference issues
+    if frame is not None:
+        try:
+            frame = frame.copy()            
+        except Exception as e:
+            print(f"ERROR: Failed to copy frame: {str(e)}")
+            frame = None
+    
+    if frame is None or frame.size == 0 or len(frame.shape) < 2:
+        print(f"ERROR: Invalid frame received for {camera_id}, task {task_uuid}")
+        return ""
+        
+    root_path = task_settings.get("root_path", "./assets/saved_videos")
+    file_type = task_settings.get("file_type", "mp4")
+    resolution = task_settings.get("resolution", "Lowest")
+    interval = int(task_settings.get("interval", 3600))
+    interval_units = task_settings.get("interval_units", "Seconds")
+    if interval_units != "Seconds":
+        interval *= time_units.get(interval_units, 1)
+    duration = int(task_settings.get("clip_length", interval))
+    duration_units = task_settings.get("clip_length_units", "Seconds")
+    if duration_units != "Seconds":
+        duration *= time_units.get(duration_units, 1)
+    start_time = task_settings.get("start_time", datetime.now(timezone.utc).isoformat())
+    timestamp = datetime.now(timezone.utc).isoformat()
+   
+    #Get frame shape for initialization
+    frame_shape = frame.shape
+   
+    saved_file_path = ""  #Initialize return value
+    
+    #If idling is not set up, initialize it
+    if camera_id not in capture_sessions.idling:
+        capture_sessions.idling[camera_id] = {}
+    if task_uuid not in capture_sessions.idling[camera_id]:
+        capture_sessions.idling[camera_id][task_uuid] = { 'last_interval_position': 0 }
+    
+    should_start = capture_sessions.should_start_capture(camera_id, task_uuid, timestamp, start_time, interval)
+
+    #Check if we have an active session and it should end
+    active_session = camera_id in capture_sessions.sessions and task_uuid in capture_sessions.sessions[camera_id]
+    should_end = active_session and capture_sessions.should_end_session(camera_id, task_uuid, timestamp, duration)
+    
+    if should_end or should_start:
```

Comment:
these 2 ifs can be combined into 1 - `if should_end and active_session` and you can move line 251 before 263